source/adios2/core/IO.cpp

```diff
@@ -0,0 +1,279 @@
+            ? true
+            : false;
+
+    const bool isDefaultReader =
+        m_EngineType.empty() &&
+                (openMode == OpenMode::Read || openMode == OpenMode::r ||
+                 openMode == OpenMode::ReadWrite || openMode == OpenMode::rw)
+            ? true
+            : false;
+
+    if (isDefaultWriter || m_EngineType == "BPFileWriter")
+    {
+        engine = std::make_shared<BPFileWriter>(*this, name, openMode, mpiComm);
+    }
+    else if (isDefaultReader || m_EngineType == "BPFileReader")
+    {
+        // engine = std::make_shared<BPFileReader>(*this, name, openMode,
+        // mpiComm);
+    }
+    else if (m_EngineType == "DataManWriter")
+    {
+#ifdef ADIOS2_HAVE_DATAMAN
+        engine =
+            std::make_shared<DataManWriter>(*this, name, openMode, mpiComm);
+#else
+        throw std::invalid_argument(
+            "ERROR: this version didn't compile with "
+            "DataMan library, can't Open DataManWriter\n");
+#endif
+    }
+    else if (m_EngineType == "DataManReader")
+    {
+#ifdef ADIOS2_HAVE_DATAMAN
+        engine =
+            std::make_shared<DataManReader>(*this, name, openMode, mpiComm);
+#else
+        throw std::invalid_argument(
+            "ERROR: this version didn't compile with "
+            "DataMan library, can't Open DataManReader\n");
+#endif
+    }
+    else if (m_EngineType == "ADIOS1Writer")
+    {
+#ifdef ADIOS2_HAVE_ADIOS1
+        engine = std::make_shared<ADIOS1Writer>(*this, name, openMode, mpiComm);
+#else
+        throw std::invalid_argument(
+            "ERROR: this version didn't compile with ADIOS "
+            "1.x library, can't Open ADIOS1Writer\n");
+#endif
+    }
+    else if (m_EngineType == "ADIOS1Reader")
+    {
+#ifdef ADIOS2_HAVE_ADIOS1
+        engine = std::make_shared<ADIOS1Reader>(*this, name, openMode, mpiComm);
+#else
+        throw std::invalid_argument(
+            "ERROR: this version didn't compile with ADIOS "
+            "1.x library, can't Open ADIOS1Reader\n");
+#endif
+    }
+    else if (m_EngineType == "HDF5Writer")
+    {
+#ifdef ADIOS2_HAVE_HDF5
+        engine = std::make_shared<HDF5WriterP>(*this, name, openMode, mpiComm);
+#else
+        throw std::invalid_argument("ERROR: this version didn't compile with "
+                                    "HDF5 library, can't use HDF5\n");
+#endif
+    }
+    else if (m_EngineType == "HDF5Reader")
+    {
+#ifdef ADIOS2_HAVE_HDF5
+        engine = std::make_shared<HDF5ReaderP>(*this, name, openMode, mpiComm);
+#else
+        throw std::invalid_argument("ERROR: this version didn't compile with "
+                                    "HDF5 library, can't use HDF5\n");
+#endif
+    }
+    return engine;
+}
+
+std::shared_ptr<Engine> IO::Open(const std::string &name,
+                                 const OpenMode openMode)
+{
+    return Open(name, openMode, m_MPIComm);
+}
+
+// PRIVATE Functions
+unsigned int
+IO::AddTransportParameters(const std::string type,
```

Comment:
It would also be good to have a signature that could accept an std::map.  That would allow you to bypass the vector -> map parsing and call it with `{"key1": "val1", "key2" : "val2"}` syntax.